atom/browser/net/url_request_fetch_job.cc

```diff
@@ -169,22 +170,21 @@ void URLRequestFetchJob::HeadersCompleted() {
 int URLRequestFetchJob::DataAvailable(net::IOBuffer* buffer,
                                       int num_bytes,
                                       const net::CompletionCallback& callback) {
-  // Store the data in a drainable IO buffer if pending_buffer_ is empty, i.e.
-  // there's no ReadRawData() operation waiting for IO completion.
+  // When pending_buffer_ is empty, there's no ReadRawData() operation waiting
+  // for IO completion, we have to save the parameters until the request is
+  // ready to read data.
   if (!pending_buffer_.get()) {
-    response_piper_callback_ = callback;
-    drainable_buffer_ = new net::DrainableIOBuffer(buffer, num_bytes);
+    write_buffer_ = buffer;
+    write_num_bytes_ = num_bytes;
+    write_callback_ = callback;
     return net::ERR_IO_PENDING;
   }
 
-  // pending_buffer_ is set to the IOBuffer instance provided to ReadRawData()
-  // by URLRequestJob.
-  int bytes_read = std::min(num_bytes, pending_buffer_size_);
-  memcpy(pending_buffer_->data(), buffer->data(), bytes_read);
-
-  ClearPendingBuffer();
-
+  // Write data to the pending buffer and clear them after the writing.
+  int bytes_read = BufferCopy(buffer, num_bytes,
+                              pending_buffer_.get(), pending_buffer_size_);
   ReadRawDataComplete(bytes_read);
+  ClearPendingBuffer();
```

Comment:
`ClearPendingBuffer` should be called before notifying `ReadRawDataComplete` otherwise possibility of clearing the wrong buffers. Should probably add a test case with multiple resources.